cai_causal_graph/identify_utils.py

```diff
@@ -90,46 +92,42 @@ def identify_confounders(graph: CausalGraph, node_1: NodeLike, node_2: NodeLike)
     :return: A list of all confounders between `node_1` and `node_2`.
     """
 
-    def _identify_confounders_no_checks_no_descendant_pruning(clean_graph: CausalGraph, n1: str, n2: str) -> List[str]:
-        """Private function that does not check if DAG or do descendant pruning."""
+    def _identify_confounders_no_checks_no_descendant_pruning_networkx(
+        clean_graph: networkx.DiGraph, n1: str, n2: str
+    ) -> Set[str]:
+        """
+        Private function that does not check if DAG or do descendant pruning.
 
+        This function searches for confounders using the node `n1` as the starting point. Importantly, this
+        does not produce a set of minimal confounders.
+        """
         # create a copy of the provided graph and prune the children of node 1 and node 2
-        final_graph = clean_graph.copy()
-        for child in final_graph.get_children(n1):
-            final_graph.remove_edge(source=n1, destination=child)
-        for child in final_graph.get_children(n2):
-            final_graph.remove_edge(source=n2, destination=child)
+        removed_edges = list()
+        final_graph = clean_graph
```

Comment:
Why no need to copy? The same graph is passed in to identify confounders and reverse confounders below. So is it not being modified in place and the wrong graph will be passed during the second call to `_identify_confounders_no_checks_no_descendant_pruning_networkx`?

Reply:
I think the recursive call below is fine but maybe it is better to copy the original `digraph` when called twice in the actual `identify_confounders`.

Reply:
Also answering your comment below:

Copying the graph is quite memory and time intensive, so the point here is to avoid that altogether. The way it is done now, we keep the integrity of the causal graph (i.e. add any removed edges back) within a specific recusrive call, meaning at the end of `_identify_confounders_no_checks_no_descendant_pruning_networkx` the graph is returned to its original state.

As I said - copying (even just inside `networkx` is memory and time intensive). Moreover, there is no real need for it since the new approach works.

What do you think?

Reply:
Just to give you some numbers, on a usecase I am running, without copying it takes 5s, while with it takes 15s.

Reply:
Sounds good. I was guessing memory but didn't think it would be so intensive. Thanks for the explanation.